manifest-build-tools/application/version_generator.py

```diff
@@ -0,0 +1,200 @@
+--is-official-release
+--parameter-file version.txt
+
+Because this script need to import scripts under lib.
+The script HWIMO-BUILD helps to add the scripts under lib to python path.
+
+The required parameters: 
+repo-dir
+manifest-repo-dir
+
+The optional parameters:
+is-official-release (default value is false)
+parameter-file (default value is release_version)
+"""
+import os
+import sys
+import argparse
+import datetime
+import subprocess
+from RepositoryOperator import RepoOperator
+from common import *
+
+class VersionGenerator(object):
+    def __init__(self, repo_dir, manifest_repo_dir):
+        """
+        :return:
+ 
+        """
+        self._repo_dir = repo_dir
+        self._manifest_repo_dir = manifest_repo_dir
+        self.repo_operator = RepoOperator()
+ 
+    def generate_small_version(self):
+        """
+        generate the small version which consists of commit date and commit hash
+        return: small version 
+        """
+
+        ts_str = self.repo_operator.get_lastest_commit_date(self._manifest_repo_dir)
+        date = datetime.datetime.utcfromtimestamp(int(ts_str)).strftime('%Y%m%d%H%M%SZ')
+        commit_id = self.repo_operator.get_lastest_commit_id(self._manifest_repo_dir)
+        version = "{date}-{commit}".format(date=date, commit=commit_id[0:7])
+        return version
+
+    def generate_big_version(self):
+        """
+        generate the big version according to changelog
+        return: big version
+        """
+        #If the repository is on-http, sync the debianstatic/on-http/ to debian before compute version
+        repo_url = self.repo_operator.get_repo_url(self._repo_dir)
+        repo_name = strip_suffix(os.path.basename(repo_url), ".git")
+        if repo_name == "on-http":
+            cmd_args = ["rsync", "-ar", "debianstatic/on-http/", "debian"]
+            proc = subprocess.Popen(cmd_args,
+                                    cwd=self._repo_dir,
+                                    stderr=subprocess.PIPE,
+                                    stdout=subprocess.PIPE,
+                                    shell=False)
+
+            (out, err) = proc.communicate()
+            if proc.returncode != 0:
+                raise RuntimeError("Failed to sync on-http/debianstatic/on-http to on-http/debian due to {0}".format(err))
+                
+        cmd_args = ["dpkg-parsechangelog", "--show-field", "Version"]
+        proc = subprocess.Popen(cmd_args,
+                                cwd=self._repo_dir,
+                                stderr=subprocess.PIPE,
+                                stdout=subprocess.PIPE,
+                                shell=False)
+
+        (out, err) = proc.communicate()
+        if proc.returncode == 0:
+            return out.strip()
+        else:
+            print "Failed to parse version in debian/changelog due to {0}".format(err)
+            return None
+
+        
+    def generate_candidate_version(self):
```

Comment:
how about call it generate_version_stage(self) ?
people may use the term "stage of development" for those words: alpha, beta, devel ,RC...
https://en.wikipedia.org/wiki/Software_release_life_cycle